drivers/adc/adc_esp32.c

```diff
@@ -0,0 +1,334 @@
+#define VALID_RESOLUTION(r) ((r) >= ADC_RESOLUTION_MIN && (r) <= ADC_RESOLUTION_MAX)
+#define INVALID_RESOLUTION(r) (!VALID_RESOLUTION(r))
+
+/* Default internal reference voltage */
+#define ADC_ESP32_DEFAULT_VREF_INTERNAL (1100)
+
+struct adc_esp32_conf {
+	adc_unit_t unit;
+	uint8_t channel_count;
+};
+
+struct adc_esp32_data {
+	adc_atten_t attenuation[ADC_CHANNEL_MAX];
+	uint8_t resolution[ADC_CHANNEL_MAX];
+	esp_adc_cal_characteristics_t chars[ADC_CHANNEL_MAX];
+	uint16_t meas_ref_internal;
+	uint16_t *buffer;
+	uint16_t *buffer_repeat;
+	bool calibrate;
+};
+
+/* Convert zephyr,gain property to the ESP32 attenuation */
+static inline int gain_to_atten(enum adc_gain gain, adc_atten_t *atten)
+{
+	switch (gain) {
+	case ADC_GAIN_1:
+		*atten = ADC_ATTEN_DB_0;
+		break;
+	case ADC_GAIN_4_5:
+		*atten = ADC_ATTEN_DB_2_5;
+		break;
+	case ADC_GAIN_1_2:
+		*atten = ADC_ATTEN_DB_6;
+		break;
+	case ADC_GAIN_1_4:
+		*atten = ADC_ATTEN_DB_11;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+	return 0;
+}
+
+/* Convert voltage by inverted attenuation to support zephyr gain values */
+static void atten_to_gain(adc_atten_t atten, uint32_t *val_mv)
+{
+	if (!val_mv) {
+		return;
+	}
+	switch (atten) {
+	case ADC_ATTEN_DB_2_5:
+		*val_mv = (*val_mv * 4) / 5; /* 1/ADC_GAIN_4_5 */
+		break;
+	case ADC_ATTEN_DB_6:
+		*val_mv = *val_mv >> 1; /* 1/ADC_GAIN_1_2 */
+		break;
+	case ADC_ATTEN_DB_11:
+		*val_mv = *val_mv / 4; /* 1/ADC_GAIN_1_4 */
+		break;
+	case ADC_ATTEN_DB_0: /* 1/ADC_GAIN_1 */
+	default:
+		break;
+	}
+}
+
+static bool adc_calibration_init(const struct device *dev)
+{
+	struct adc_esp32_data *data = dev->data;
+
+	switch (esp_adc_cal_check_efuse(ADC_CALI_SCHEME)) {
+	case ESP_ERR_NOT_SUPPORTED:
+		LOG_WRN("Skip software calibration - Not supported!");
+		break;
+	case ESP_ERR_INVALID_VERSION:
+		LOG_WRN("Skip software calibration - Invalid version!");
+		break;
+	case ESP_OK:
+		LOG_DBG("Software calibration possible");
+		return true;
+	default:
+		LOG_ERR("Invalid arg");
+		break;
+	}
+	return false;
+}
+
+static int adc_esp32_read(const struct device *dev, const struct adc_sequence *seq)
+{
+	const struct adc_esp32_conf *conf = dev->config;
+	struct adc_esp32_data *data = dev->data;
+	int reading;
+	uint32_t cal, cal_mv;
+
+	uint8_t channel_id = find_lsb_set(seq->channels) - 1;
+
+	if (seq->buffer_size < 2) {
+		LOG_ERR("Sequence buffer space too low '%d'", seq->buffer_size);
+		return -ENOMEM;
+	}
+
+	if (seq->channels > BIT(channel_id)) {
+		LOG_ERR("Multi-channel readings not supported");
+		return -ENOTSUP;
+	}
+
+	if (INVALID_RESOLUTION(seq->resolution)) {
+		LOG_ERR("unsupported resolution (%d)", seq->resolution);
+		return -ENOTSUP;
+	}
+
+	if (seq->calibrate) {
+		/* TODO: Does this mean actual Vref measurement on selected GPIO ?*/
+		LOG_ERR("calibration is not supported");
+		return -ENOTSUP;
+	}
+
+	data->resolution[channel_id] = seq->resolution;
+
+#if CONFIG_IDF_TARGET_ESP32C3
```

Comment:
Although `CONFIG_IDF_TARGET_ESP32C3` is a valid item, I recommend converting all `CONFIG_IDF_TARGET_ESP32xx` instances to `CONFIG_SOC_ESP32xx` to make it "Zephyr-like".